### ISSUE TITLE
[Scheduler] Using before, after or onFailure methods causes unhandled exception when not configured

### ISSUE THREAD

**Issue by CoalaJoe** [symfony/symfony#53722]
### Symfony version(s) affected

6.4

### Description

According to the documentation at: 
https://symfony.com/doc/current/scheduler.html#a-strategic-event-handling
<img width="756" alt="image" src="https://github.com/symfony/symfony/assets/5689499/a0d859b8-e005-4e73-834b-9543fd361882">

Using those functions like that does not work and results in a 500 error
```
{"message":"Uncaught Error: Call to a member function addListener() on null","context":{"exception":{"class":"Error","message":"Call to a member function addListener() on null","code":0,"file":"/app/vendor/symfony/scheduler/Schedule.php:144"}},"channel":"php","extra":{},"severity":"CRITICAL","time":"2024-02-01T13:22:04.559+01:00"}
```

### How to reproduce

```
     // ...
    public function getSchedule(): Schedule
    {
        $this->schedule ??= (new Schedule())
            ->with(...$this->getGeneralMessages(), ...$this->getProductionMessages())
            ->before(function (PreRunEvent $event) {
                $this->logger?->info('Before Event');
            })
            ->after(function (PostRunEvent $event) {
                $this->logger?->info('After Event');
            })
            ->onFailure(function (FailureEvent $event) {
                $this->logger?->info('Failure Event');
                $this->logger?->info($event->getError()->getMessage());
            })
            ->lock($this->lockFactory->createLock('cron-cache'))
            ->stateful($this->cache);

        return $this->schedule;
    }
```

### Possible Solution

Either update documentaiton, add error handling and throw our own exception or fail silently and log an error?

**Comment by kl3sk** [comment]
I face the same problem and manage to fix it by injecting the dispatcher on Scheduler contruct

```php
    public function __construct(
        private readonly CacheInterface $cache,
        private readonly EventDispatcherInterface $dispatcher,
    )
    {
    }

    public function getSchedule(): Schedule
    {
        return $this->schedule ??= (new Schedule($this->dispatcher))
            ->before(function (PreRunEvent $event) {
...
```

**Comment by fabpot** [comment]
Code fix is here: #57282
Doc fix is here: https://github.com/symfony/symfony-docs/pull/19935
